bundles/io/org.eclipse.smarthome.io.rest/src/main/java/org/eclipse/smarthome/io/rest/Stream2JSONInputStream.java

```diff
@@ -0,0 +1,116 @@
+ * from the top level elements of the stream. For best performance a flattened stream should be provided. Otherwise a
+ * nested collections JSON representation will be fully transformed into memory.
+ *
+ * @author Henning Treu - initial contribution
+ *
+ */
+public class Stream2JSONInputStream extends InputStream {
+
+    private final Iterator<String> iterator;
+
+    private InputStream jsonElementStream;
+
+    private boolean firstIteratorElement;
+
+    private Gson GSON = new GsonBuilder().create();
+
+    /**
+     * Creates a new {@link Stream2JSONInputStream} backed by the given {@link Stream} source.
+     *
+     * @param source the {@link Stream} backing this input stream. Must not be null.
+     *
+     * @throws IllegalArgumentException in case the source is null.
+     */
+    public Stream2JSONInputStream(Stream<?> source) {
+        if (source == null) {
+            throw new IllegalArgumentException("The source must not be null!");
+        }
+
+        iterator = source.map(e -> GSON.toJson(e)).iterator();
+        jsonElementStream = IOUtils.toInputStream("");
+        firstIteratorElement = true;
+    }
+
+    @Override
+    public int read() throws IOException {
+        int result = jsonElementStream.read();
+
+        if (result == -1) { // the current JSON element was completely streamed
+            if (finished()) { // we are done streaming the collection
+                return -1;
+            }
+
+            fillBuffer(); // get the next element into the jsonInputStream
+            result = jsonElementStream.read();
+        }
+
+        return result;
+    }
+
+    @Override
+    public void close() throws IOException {
+        closeJSONStream();
+        super.close();
+    }
+
+    private void fillBuffer() throws IOException {
+        String prefix;
+        if (firstIteratorElement) {
+            prefix = "[";
+            firstIteratorElement = false;
+        } else {
+            prefix = ",";
+        }
+
+        String entity = iterator.hasNext() ? iterator.next() : "";
+
+        String postfix = "";
+        if (!iterator.hasNext()) {
+            postfix = "]";
+        }
+
+        try {
+            closeJSONStream();
+            jsonElementStream = IOUtils.toInputStream(prefix + entity + postfix, StandardCharsets.UTF_8.name());
+        } catch (IOException e) {
+            // IOException is thrown for invalid encoding. This will never happen with StandardCharsets.UTF_8.
+        }
+    }
+
+    private boolean finished() {
+        return !firstIteratorElement && !iterator.hasNext();
+    }
+
+    private void closeJSONStream() throws IOException {
+        if (jsonElementStream != null) {
+            try {
+                jsonElementStream.close();
+            } catch (IOException e) {
+                throw new IOException("Error closing the underlying JSON stream.", e);
```

Comment:
Do you really need to encapsulate the `IOException` into another `IOException` only to add a new message?
Shouldn't it be enough to have a look at the stack trace if someone needs to further identify error?